### PR TITLE
Implement rct

### LINKED ISSUE #63: Implement randomized control trial in api
Plan:
- First create a prediction per day (like now)
- Then based on prediction score bins (using quantile bins, for example 10) and clinic create stratified randomization in control and treatment groups
- Save group to the prediction in the database
- exclude control group from dashboard

### PR DESCRIPTION
**Pull request by WoltersEric**
closes #63

### CODE REVIEW

File: src/noshow/api/app_helpers.py [hunk 1]
```diff
@@ -69,3 +70,50 @@ def fix_outdated_appointments(
...
+    Raises
+    ------
+    ValueError
+        If the predictions DataFrame is empty.
+    """
+    if predictions.empty:
+        raise ValueError("The predictions DataFrame is empty.")
+
+    # Create prediction score bins using quantile bins, for example 10
+    predictions["score_bin"] = pd.qcut(predictions["prediction"], q=10)
```
Comment (rubenpeters91) [inline src/noshow/api/app_helpers.py:98]: Maybe make the q-parameter a optional parameter of the function (with default value 10), that could also prevent the unit test error

File: src/noshow/api/app_helpers.py [hunk 2]
```diff
@@ -69,3 +70,50 @@ def fix_outdated_appointments(
...
+
+    # sort predictions by hoofdagenda and score_bin
```
Comment (rubenpeters91) [inline src/noshow/api/app_helpers.py:100]: This comment is not correct, also doesn't really add anything

File: src/noshow/api/app_helpers.py [hunk 3]
```diff
@@ -69,3 +70,50 @@ def fix_outdated_appointments(
...
+    predictions = predictions.sort_values(["prediction"])
+
+    # Pre-process data to handle duplicates
```
Comment (rubenpeters91) [inline src/noshow/api/app_helpers.py:103]: This comment also doens't explain why, a better comment would be: `only keep top prediction per patient for treatment/control split`

File: src/noshow/api/app_helpers.py [hunk 4]
```diff
@@ -69,3 +70,50 @@ def fix_outdated_appointments(
...
+    deduplicated = predictions.drop_duplicates(subset="pseudo_id", keep="first")
+
+    # Create stratified randomization in control and treatment groups
+    deduplicated = deduplicated.sort_values(["hoofdagenda", "prediction"])
+    deduplicated["treatment_group"] = deduplicated.groupby(
+        ["hoofdagenda", "score_bin"], observed=True
+    )["prediction"].transform(lambda x: np.arange(len(x)) % 2)
+    # Merge back the treatment_group assignments to the original DataFrame
```
Comment (rubenpeters91) [inline src/noshow/api/app_helpers.py:111]: This also just explains what we can already read in the code below. copilot? ;)

File: tests/test_rct.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    """Fixture to create a sample DataFrame for testing."""
+    data = {
+        "prediction": [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
+        "hoofdagenda": ["A", "A", "B", "B", "C", "C", "D", "D", "E", "E"],
+        "pseudo_id": [1, 2, 3, 1, 3, 2, 4, 1, 2, 4],
+    }
+    return pd.DataFrame(data)
+
+
+class TestRCT:
```
Comment (rubenpeters91) [inline tests/test_rct.py:26]: I don't think it's neccessary to group the tests in a class right?

**Review by rubenpeters91 — APPROVED** [review]
Looks good, just some small comments. And the unit test is still failing, probably because with the current test dataset there are not enough predictions to create 10 quantile bins

File: src/noshow/api/app_helpers.py (hunk 1 above)
Comment (WoltersEric) [reply src/noshow/api/app_helpers.py:98]: good one

Comment (WoltersEric) [reply src/noshow/api/app_helpers.py]: Turns out it was another issue. The mock model only created predictions of 0.0 so I changed it to created random predictions instead

**Review by rubenpeters91 — APPROVED** [review]
Looks good

**Pull request merged**